scheduler/src/cook/mesos/share.clj

```diff
@@ -44,30 +44,59 @@
                    [[:db.fn/retractEntity resource]])
       (log/warn "Resource" type "for user" user "does not exist, could not retract"))))
 
-(defn- get-share-by-type
-  [db type user]
-  (let [type (resource-type->datomic-resource-type type)
-        query '[:find ?a
-                :in $ ?u ?t
-                :where
-                [?e :share/user ?u]
-                [?e :share/resource ?r]
-                [?r :resource/type ?t]
-                [?r :resource/amount ?a]]]
-    (or (-> (q query db user type) ffirst)
-        (-> (q query db default-user type) ffirst)
-        (Double/MAX_VALUE))))
+(let [query '[:find ?a
+              :in $ ?u ?t
+              :where
+              [?e :share/user ?u]
+              [?e :share/resource ?r]
+              [?r :resource/type ?t]
+              [?r :resource/amount ?a]]]
+  (defn- get-share-by-type
+    "Query a user's pre-defined share by type (e.g. :cpus, :mem, :gpus).
+
+     If a user's pre-defined share is NOT defined, return either:
+     1. the explicitly provided default value, or
+     2. the share for the \"default\" user. If there is NO \"default\"
+        value for a specific type, return Double.MAX_VALUE."
+    ([db type user]
+     (let [datomic-resource-type (resource-type->datomic-resource-type type)]
+       (or (ffirst (q query db user datomic-resource-type))
+           (get-share-by-type db type default-user Double/MAX_VALUE))))
+    ([db type user default]
+     (let [datomic-resource-type (resource-type->datomic-resource-type type)]
```

Comment:
Thanks for pointing this out. Made two recursive calls and used an `or`. This further allowed privatizing the query. Sweet!